Remove non-standard `percentage` value from `-webkit-line-clamp`

Ahmad-S792 opened the pull request:
#### 8cbbfc9b154d2063cf96f04b40ae7e437fac9241
<pre>
Remove non-standard `percentage` value from `-webkit-line-clamp`
<a href="https://bugs.webkit.org/show_bug.cgi?id=298000">https://bugs.webkit.org/show_bug.cgi?id=298000</a>
<a href="https://rdar.apple.com/159808285">rdar://159808285</a>

Reviewed by Alan Baradlay.

This patch removes non-standard percentage value from `-webkit-line-clamp`,
which is not present in new standard for `line-clamp` plus this was also
dropped from Blink in 2018 [1] due to lack of usage and it also has quirky
undefined behavior different from web developer expectations in percentage
distribution between lines.

[1] <a href="https://chromium.googlesource.com/chromium/src.git/+/7dc2379d58e26905dd77c56617d0baf9972581eb">https://chromium.googlesource.com/chromium/src.git/+/7dc2379d58e26905dd77c56617d0baf9972581eb</a>

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::hasClampingAndNoFlexing const):
(WebCore::RenderDeprecatedFlexibleBox::applyLineClamp):
(WebCore::lineCountFor): Deleted.
* Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.cpp: Removed.
* Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.h:
(WebCore::Style::WebkitLineClamp::WebkitLineClamp): Deleted.
(WebCore::Style::WebkitLineClamp::isNone const): Deleted.
(WebCore::Style::WebkitLineClamp::isPercentage const): Deleted.
(WebCore::Style::WebkitLineClamp::isInteger const): Deleted.
(WebCore::Style::WebkitLineClamp::tryPercentage const): Deleted.
(WebCore::Style::WebkitLineClamp::tryInteger const): Deleted.
(WebCore::Style::WebkitLineClamp::holdsAlternative const): Deleted.
(WebCore::Style::WebkitLineClamp::switchOn const): Deleted.
(WebCore::Style::WebkitLineClamp::valueForTextAutosizingHash const): Deleted.

&gt; Tests:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value-expected.txt:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value.html:
* LayoutTests/fast/overflow/line-clamp.html:
* LayoutTests/platform/mac/fast/overflow/line-clamp-expected.txt:
* LayoutTests/platform/ios/fast/overflow/line-clamp-expected.txt:
* LayoutTests/platform/glib/fast/overflow/line-clamp-expected.txt:

Canonical link: <a href="https://commits.webkit.org/318858@main">https://commits.webkit.org/318858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef9f54eab21a8477bd2714f1822583b612f1f7c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/179531 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/53470 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/47268 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/189363 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/143840 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cc6618ee-029d-4b11-8c9c-425f9daa4e9e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/181394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/54764 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/53181 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/140007 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/96860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/35b8ce50-c82a-4464-b09e-99168ed783e7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/182488 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/43619 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/160751 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/120460 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1d0828a9-4b25-444c-9300-c24a6f7499f7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/42289 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/40613 "Passed tests") | | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/152690 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/40263 "Passed tests") | [✅ 🛠 gtk3-gcc](https://ews-build.webkit.org/#/builders/284/builds/817 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/46076 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/44861 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/191584 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/53140 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/160268 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/149270 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/40028 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/53821 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/36887 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/149017 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/43679 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/126093 "Built successfully") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/120997 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/52338 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/15249 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/51723 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/51964 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/51784 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.h:
```diff
@@ -30,32 +30,24 @@
+// <-webkit-line-clamp> = none | <integer [1,inf]>
 // NOTE: CSS Overflow 4 re-defines `-webkit-line-clamp` as a shorthand of 'max-lines', 'block-ellipsis', and 'continue'.
 // https://drafts.csswg.org/css-overflow-4/#webkit-line-clamp
 struct WebkitLineClamp {
```
This should use the ValueOrKeyword helper base class now that it is just integer and none.

weinig replied:
To be more clear, since this is not something a lot of people touch, the new struct would look like this:

```c++
struct WebkitLineClamp : ValueOrKeyword<Integer<CSS::Range{1, CSS::Range::infinity}>, CSS::Keyword::None> {
    using Base::Base;
    using Integer = Base::Value;

    constexpr bool isNone() const { return isKeyword(); }
    constexpr bool isInteger() const { return isValue(); }
    std::optional<Integer> tryInteger() const { return tryValue(); }

    unsigned valueForTextAutosizingHash() const
    {
        return switchOn(
            [](const CSS::Keyword::None&) { return 0; },
            [](const Integer& integer) { return static_cast<unsigned>(integer.value); }
        );
    }
};
```

And then you can remove the lines:

```
// MARK: - Conversion

template<> struct CSSValueConversion<WebkitLineClamp> { auto operator()(BuilderState&, const CSSValue&) -> WebkitLineClamp; };
```

and remove the entire cpp file.

Ahmad-S792 committed:
Remove non-standard `percentage` value from `-webkit-line-clamp`
https://bugs.webkit.org/show_bug.cgi?id=298000
rdar://159808285

Reviewed by NOBODY (OOPS!).

This patch removes non-standard percentage value from `-webkit-line-clamp`,
which is not present in new standard for `line-clamp` plus this was also
dropped from Blink in 2018 [1] due to lack of usage and it also has quirky
undefined behavior different from web developer expectations in percentage
distribution between lines.

[1] https://chromium.googlesource.com/chromium/src.git/+/7dc2379d58e26905dd77c56617d0baf9972581eb

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::hasClampingAndNoFlexing const):
(WebCore::RenderDeprecatedFlexibleBox::applyLineClamp):
(WebCore::lineCountFor): Deleted.
* Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.cpp: Removed.
* Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.h:
(WebCore::Style::WebkitLineClamp::WebkitLineClamp): Deleted.
(WebCore::Style::WebkitLineClamp::isNone const): Deleted.
(WebCore::Style::WebkitLineClamp::isPercentage const): Deleted.
(WebCore::Style::WebkitLineClamp::isInteger const): Deleted.
(WebCore::Style::WebkitLineClamp::tryPercentage const): Deleted.
(WebCore::Style::WebkitLineClamp::tryInteger const): Deleted.
(WebCore::Style::WebkitLineClamp::holdsAlternative const): Deleted.
(WebCore::Style::WebkitLineClamp::switchOn const): Deleted.
(WebCore::Style::WebkitLineClamp::valueForTextAutosizingHash const): Deleted.

> Tests:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value-expected.txt:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value.html:
* LayoutTests/fast/overflow/line-clamp.html:
* LayoutTests/platform/mac/fast/overflow/line-clamp-expected.txt:
* LayoutTests/platform/ios/fast/overflow/line-clamp-expected.txt:
* LayoutTests/platform/glib/fast/overflow/line-clamp-expected.txt:

Ahmad-S792 commented on Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.h:
```diff
@@ -30,62 +30,26 @@
     {
-        return WTF::switchOn(m_value,
+        return switchOn(
             [](const CSS::Keyword::None&) { return std::numeric_limits<unsigned>::max(); },
```
@weinig - I tried to use `return 0;` here but it was not compiling.

Ahmad-S792 replied:
```/WebCore/css/values/CSSValueAggregates.h:582:9: error: 
      'decltype(auto)' in return type deduced as 'unsigned int' here but deduced as 'int' in earlier return statement
  582 |         return visitor(*m_value);
  ```

nullhook replied:
your first lambda is being deduced as int `return 0;` second as unsigned.

you can can be explicit `[](const CSS::Keyword::None&) { return 0u; }`

Ahmad-S792 replied:
@nullhook - Thanks, it worked.

weinig commented on Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.h:
```diff
@@ -31,62 +31,26 @@
-struct WebkitLineClamp {
-    using Percentage = Style::Percentage<CSS::Nonnegative>;
-    using Integer = Style::Integer<CSS::Range{1, CSS::Range::infinity}>;
+struct WebkitLineClamp : ValueOrKeyword<Integer<CSS::Range { 1, CSS::Range::infinity }>, CSS::Keyword::None> {
```
```suggestion
struct WebkitLineClamp : ValueOrKeyword<Integer<CSS::Positive>, CSS::Keyword::None> {
```

Ahmad-S792 replied:
@weinig - Done. Thanks!

Ahmad-S792 committed:
Remove non-standard `percentage` value from `-webkit-line-clamp`
https://bugs.webkit.org/show_bug.cgi?id=298000
rdar://159808285

Reviewed by Alan Baradlay.

This patch removes non-standard percentage value from `-webkit-line-clamp`,
which is not present in new standard for `line-clamp` plus this was also
dropped from Blink in 2018 [1] due to lack of usage and it also has quirky
undefined behavior different from web developer expectations in percentage
distribution between lines.

[1] https://chromium.googlesource.com/chromium/src.git/+/7dc2379d58e26905dd77c56617d0baf9972581eb

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::hasClampingAndNoFlexing const):
(WebCore::RenderDeprecatedFlexibleBox::applyLineClamp):
(WebCore::lineCountFor): Deleted.
* Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.cpp: Removed.
* Source/WebCore/style/values/non-standard/StyleWebKitLineClamp.h:
(WebCore::Style::WebkitLineClamp::WebkitLineClamp): Deleted.
(WebCore::Style::WebkitLineClamp::isNone const): Deleted.
(WebCore::Style::WebkitLineClamp::isPercentage const): Deleted.
(WebCore::Style::WebkitLineClamp::isInteger const): Deleted.
(WebCore::Style::WebkitLineClamp::tryPercentage const): Deleted.
(WebCore::Style::WebkitLineClamp::tryInteger const): Deleted.
(WebCore::Style::WebkitLineClamp::holdsAlternative const): Deleted.
(WebCore::Style::WebkitLineClamp::switchOn const): Deleted.
(WebCore::Style::WebkitLineClamp::valueForTextAutosizingHash const): Deleted.

> Tests:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value-expected.txt:
* LayoutTests/fast/css/webkit-line-clamp-calculated-value.html:
* LayoutTests/fast/overflow/line-clamp.html:
* LayoutTests/platform/mac/fast/overflow/line-clamp-expected.txt:
* LayoutTests/platform/ios/fast/overflow/line-clamp-expected.txt:
* LayoutTests/platform/glib/fast/overflow/line-clamp-expected.txt:

Canonical link: https://commits.webkit.org/318858@main
